### PR TITLE
Add support for Llama2, Palm, Cohere, Anthropic, Replicate, Azure Models - using litellm

### PR DESCRIPTION
**Pull request by ishaan-jaff**
This PR adds support for 50+ models with a standard I/O interface using: https://github.com/BerriAI/litellm/

`ChatLiteLLM()` is integrated into langchain and allows you to call all models using the `ChatOpenAI` I/O interface 
https://python.langchain.com/docs/integrations/chat/litellm

Here's an example of how to use ChatLiteLLM()
```python
ChatLiteLLM(model="gpt-3.5-turbo")
ChatLiteLLM(model="claude-2", temperature=0.3)
ChatLiteLLM(model="command-nightly")
ChatLiteLLM(model="replicate/llama-2-70b-chat:2c1608e18606fad2812020dc541930f2d0495ce32eee50074220b87300bc16e1")

```

### CODE REVIEW

**Comment by ishaan-jaff** [comment]
@filip-michalsky can I get a review on this PR please😊

if this initial commit looks good happy to add more docs/ tests too

**Comment by sunyilong0** [comment]
so good

**Comment by krrishdholakia** [comment]
bump on this? @filip-michalsky

**Comment by filip-michalsky** [comment]
Need to add docs:

pip installs necessary 
- google-generativeai (version?)
- `langchain` minimum version where `litellm` was merged?

**Comment by filip-michalsky** [comment]
Just checkout out the PR locally and running to this error:

  File "/Users/filipmichalsky/Odyssey/sales_bot/SalesGPT/run.py", line 34, in <module>
    llm = ChatLiteLLM(temperature=0.2)
  File "/Users/filipmichalsky/anaconda3/envs/salesgpt/lib/python3.10/site-packages/langchain/load/serializable.py", line 74, in __init__
    super().__init__(**kwargs)
  File "pydantic/main.py", line 339, in pydantic.main.BaseModel.__init__
  File "pydantic/main.py", line 1102, in pydantic.main.validate_model
  File "/Users/filipmichalsky/anaconda3/envs/salesgpt/lib/python3.10/site-packages/langchain/chat_models/litellm.py", line 285, in validate_environment
    raise ChatLiteLLMException(
langchain.chat_models.litellm.ChatLiteLLMException: Could not import google.generativeai python package. Please install it with `pip install google-generativeai`

File: examples/streaming_generator_example.py [hunk 1]
```diff
@@ -1,14 +1,14 @@
...
-from langchain.chat_models import ChatOpenAI
+from langchain.chat_models import ChatLiteLLM
 
 with open('.env','r') as f:
     env_file = f.readlines()
 envs_dict = {key.strip("'") :value.strip("\n") for key, value in [(i.split('=')) for i in env_file]}
 os.environ['OPENAI_API_KEY'] = envs_dict['OPENAI_API_KEY']
 
-llm = ChatOpenAI(temperature=0.9)
+llm = ChatLiteLLM(temperature=0.9)
```
Comment (filip-michalsky) [inline examples/streaming_generator_example.py:11]: Does this library support `https://github.com/filip-michalsky/SalesGPT/blob/main/salesgpt/agents.py#L139`?

we are leveraging `completion_with_retry` from core langchain.

**Review by filip-michalsky — CHANGES_REQUESTED** [review]
Thanks so much for your PR!

I reviewed and I am OK with using ChatLiteLLM inside SalesGPT to give users easy access to different backend LLMs.

We need a few quality improvements:

Need to add reproducibility instructions:

- Minimum library versions to `requirements.txt`

I am not convinced this will work for the streaming generator function per my comment unless you are mimicking core `langchain` `completion_with_retry` signature.

I was unable to run `python run.py` example with this library.

Please update tests as well.

**Comment by ishaan-jaff** [comment]
@filip-michalsky i tried our langchain integration on a notebook and it worked fine - there was no dependency for 

notebook: https://colab.research.google.com/drive/1IqBFz5YXIoa8XahPb3qslmg3UoWVMUR_#scrollTo=TahkCtlmCD65

**Comment by ishaan-jaff** [comment]
here's what's remaining for this PR:
@filip-michalsky  lmk if anything is missing

- verify streaming works 
- update requirements.txt 
- Add troubleshooting to readme incase users see google-generativeai error 
- verify completion_with_retry works with ChatLiteLLM

**Comment by ishaan-jaff** [comment]
as for updating requirements.txt, i'm not sure we should add `google-cloud-aiplatform` as a dependency to this project. We should only add instructions to `pip install google-cloud-aiplatform` if users are trying VertexAI LLMs

**Comment by ishaan-jaff** [comment]
Screenshot 2023-09-02 at 3 57 28 PM

I can confirm the current fixes work locally when running run.py

**Comment by ishaan-jaff** [comment]
Can confirm it works with your streaming generator too:
Screenshot 2023-09-02 at 4 06 53 PM

**Comment by ishaan-jaff** [comment]
Can confirm we support completion_with_retry:
https://github.com/langchain-ai/langchain/blob/0e4c5dd1767b53e212d63bc81cceb7270178c322/libs/langchain/langchain/chat_models/litellm.py#L267

**Comment by filip-michalsky** [comment]
> @filip-michalsky i tried our langchain integration on a notebook and it worked fine - there was no dependency for google-generativeai
> 
> notebook: https://colab.research.google.com/drive/1IqBFz5YXIoa8XahPb3qslmg3UoWVMUR_#scrollTo=TahkCtlmCD65

Pls give me access so I can see the notebook :-)

**Comment by filip-michalsky** [comment]
> as for updating requirements.txt, i'm not sure we should add `google-cloud-aiplatform` as a dependency to this project. We should only add instructions to `pip install google-cloud-aiplatform` if users are trying VertexAI LLMs

Please add to `requirements.txt`
litellm>=

I have verified this version runs both in `run.py` and the streaming example. I also don't run into the GCP api dependency issue and agreed it should not be included in the SalesGPT project.

**Comment by filip-michalsky** [comment]
> here's what's remaining for this PR: @filip-michalsky lmk if anything is missing
> 
> * verify it works when running run.py (Done)
> * verify streaming works (Done)
> * update requirements.txt (waiting for @filip-michalsky for feedback on this)
> * Add troubleshooting to readme incase users see google-generativeai error (waiting for @filip-michalsky for feedback on this)
> * verify completion_with_retry works with ChatLiteLLM (Done)

Thank you!

It seems that you don't support asynchronous execution yet? 

I updated [this line](https://github.com/filip-michalsky/SalesGPT/pull/36/files#diff-cadc3583be4ac99d77c9eb47ad3352a1389db9caa55a5728e9a9e0e8a5a2aac0R121) in `test_salesgpt.py` with ChatLiteLLM and got the following error:


FAILED tests/test_salesgpt.py::TestSalesGPT::test_valid_async_inference_stream - AttributeError: module 'litellm' has no attribute 'acreate'

**Comment by filip-michalsky** [comment]
@ishaan-jaff thanks a lot for the PR update. Looks good from all the above, just need to add support for `acompletion_with_retry` to be fully compatible with current SalesGPT agent.

What is your ETA on supporting this if it's on the roadmap?

https://github.com/langchain-ai/langchain/blob/0e4c5dd1767b53e212d63bc81cceb7270178c322/libs/langchain/langchain/chat_models/openai.py#L79

Thank you!!

**Comment by krrishdholakia** [comment]
Hey @filip-michalsky will work on getting this done today

**Comment by filip-michalsky** [comment]
@krrishdholakia sounds good!

**Comment by luolin-ai** [comment]
> This PR adds support for 50+ models with a standard I/O interface using: https://github.com/BerriAI/litellm/
> 
> `ChatLiteLLM()` is integrated into langchain and allows you to call all models using the `ChatOpenAI` I/O interface https://python.langchain.com/docs/integrations/chat/litellm
> 
> Here's an example of how to use ChatLiteLLM()
> 
> ```python
> ChatLiteLLM(model="gpt-3.5-turbo")
> ChatLiteLLM(model="claude-2", temperature=0.3)
> ChatLiteLLM(model="command-nightly")
> ChatLiteLLM(model="replicate/llama-2-70b-chat:2c1608e18606fad2812020dc541930f2d0495ce32eee50074220b87300bc16e1")
> ```

以实现

**Comment by filip-michalsky** [comment]
@ishaan-jaff @krrishdholakia looks like async streaming is supported now - @ishaan-jaff I am happy to push to your fork to update the async generator. Also I find it super neat you can a/b test models. For me, I would love to a/b test all the LLMs supported on a given use case and benchmark for latency

**Comment by krrishdholakia** [comment]
Do you mean a/b test on a local script / jupyter notebook or in production? 

For local testing - https://docs.litellm.ai/docs/tutorials/compare_llms
For production - https://docs.litellm.ai/docs/tutorials/ab_test_llms

**Comment by filip-michalsky** [comment]
@krrishdholakia locally or in staging

**Comment by krrishdholakia** [comment]
Oh - why run this in staging?

**Comment by filip-michalsky** [comment]
@krrishdholakia network latency replication -duplicate prod. This is getting off-topic from the PR, happy to chat separately. @ishaan-jaff  lmk on the async streaming generator, happy to push to your fork (not sure of the workflow of pushing to someone else's fork of a repo though, have not done this before)

**Comment by ishaan-jaff** [comment]
@filip-michalsky updated requirements.txt to add litellm 

Pending tasks:
- looks like ChatLiteLLM failed for `acompletion_with_retry`. @filip-michalsky do you want to address that in this PR too? 
(since this PR does not edit https://github.com/langchain-ai/langchain/blob/0e4c5dd1767b53e212d63bc81cceb7270178c322/libs/langchain/langchain/chat_models/openai.py#L79)

If it makes sense happy to address that in the next PR

**Comment by filip-michalsky** [comment]
@ishaan-jaff 

SalesGPT actually has it's own `acompletion_with_retry` which can be updated inpdependent of `langchain`
https://github.com/filip-michalsky/SalesGPT/blob/main/salesgpt/agents.py#L185

I just tried this:

```
from litellm import acompletion

    async def acompletion_with_retry(self, llm: Any, **kwargs: Any) -> Any:
        """Use tenacity to retry the async completion call."""
        retry_decorator = _create_retry_decorator(llm)

        @retry_decorator
        async def _completion_with_retry(**kwargs: Any) -> Any:
            # Use OpenAI's async api https://github.com/openai/openai-python#async-api
            return await acompletion(**kwargs)

        return await _completion_with_retry(**kwargs)
```
This does not pass the async test since the object returned from litellm acompletion method with stream=True is *not* an async generator.

I feel we need to address this in this PR and hopefully will improve your `litellm` codebase as well. (When combining streaming with async, I would expect the returning object is an asynchronous generator).

Thank you!

**Comment by ishaan-jaff** [comment]
@filip-michalsky sounds good - tracking this here: https://github.com/BerriAI/litellm/issues/292 

will aim to have this done in the next 12 hours 

Agree- we should fix this in this PR and it will def make LiteLLM better too

**Comment by ishaan-jaff** [comment]
confirming tests pass on latest litellm version + with using litellm.acompletion 

Screenshot 2023-09-07 at 2 18 25 PM

File: requirements.txt [hunk 2]
```diff
@@ -1,4 +1,4 @@
-langchain>=0.0.205
+langchain>=0.0.281
```
Comment (ishaan-jaff) [inline requirements.txt:1]: 0.0.281 has the latest LiteLLM updates to langchain: 
https://github.com/langchain-ai/langchain/releases

File: requirements.txt [hunk 3]
```diff
@@ -7,4 +7,5 @@ isort>=5.12.0
 pytest>=7.3.2
 chromadb>=0.3.29
 tiktoken>=0.4.0
-pytest-asyncio
\ No newline at end of file
+pytest-asyncio
+litellm>=0.1.558
```
Comment (ishaan-jaff) [inline requirements.txt:11]: version with support for returning async generators

**Comment by ishaan-jaff** [comment]
done @filip-michalsky 

lmk if any other changes needed

**Comment by filip-michalsky** [comment]
@ishaan-jaff amazing! I see you did not update the test, but when I did, it passes now. I will just send a quick new PR to update test/changelog etc but you are good to go :) would love to have SalesGPT as one of the projects using LiteLLM on your docs :)

Thanks,

Filip

**Pull request merged**